### ISSUE TITLE
Network graph link labels

### ISSUE THREAD

**Issue by timdisher** [highcharts/highcharts#10026]
#### Expected behaviour

When used to summarize networks of clinical studies, links in network graphs can have labels indicating e.g. the number of studies in a connection (e.g. the value of the width) or optionally the names of the studies included (e.g. entered as a string: "Davis 2018 <br> Hung 2016"). 

#### Actual behaviour

Currently only markers have labels.

**Comment by sebastianbochan** [comment]
Hi @timdisher,
Thank you for contacting us. At this moment the option is available as external part of code like here:
- https://jsfiddle.net/cLspedz0/

@TorsteinHonsi  should we include it in next release of network-graph?

**Comment by timdisher** [comment]
Thank you for such a quick reply @sebastianbochan. I wonder though if there is somewhere I can read up to learn to pull the value from the array describing the connections? The idea being That A-B might describe two trials, and A-C might be two different trials. I am new to highcharts and javascript vis in general (coming from ggplot2 in R mostly) so it feels as if the learning curve just got a little steeper! 

I'm hoping to create a highcharts implementation with an htmlwidget that automates a graph that looks like the one attached. So far I've got something working based on vis.js but would prefer an implementation in highcharts. 

![image](https://user-images.githubusercontent.com/28613433/52054154-f07f2380-2531-11e9-8cb1-d0c149e7ebfc.png)

**Comment by pawelfus** [comment]
Hi @timdisher 

Welcome in JS world! :) Network graph (force directed graph) has different expected layout with your dataset, let me show you: https://jsfiddle.net/BlackLabel/xhwstejk/18/

Regarding nodes and connections, please read our docs:
- general network graph documentation: https://www.highcharts.com/docs/chart-and-series-types/networkgraph
- nodes configs: https://api.highcharts.com/highcharts/series.networkgraph.nodes
- connections configs: https://api.highcharts.com/highcharts/series.networkgraph.data

Here is a simple example how would I configure this chart: https://jsfiddle.net/BlackLabel/xhwstejk/6/ ( with dataLabels plugin: https://jsfiddle.net/BlackLabel/xhwstejk/10/ )

@TorsteinHonsi - this chart looks similar to the dependancy wheel - will that work? Also, with a bit of work, we can add a "radial" force to the FDG which will give similar layouts to the image above.

**Comment by timdisher** [comment]
You have both been a great help. Is there a resource for me to look up what the attribute options are for renderer.createElement('textPath')? You'll have to excuse me if this is a silly question  but I wasn't able to find the right combination of key terms in the API reference so I thought maybe it was a more base JS type thing?

Looking to fiddle around a bit to add some padding on the link labels so they don't overlap when the link width increases.

Also wondering if there is an easy fix to allow the label to behave like a paragraph e.g. <br> gives a carriage return. Right now adding <br> to the link label skips a line but doesn't return to the start of label.

**Comment by pawelfus** [comment]
In API go to `Classes` tab and check `SVGRenderer` class: https://api.highcharts.com/class-reference/Highcharts.SVGRenderer.html#createElement The part with `linkLabels` is not supported  (that's why this issue is marked as an enhancement ;) ) so we need to 'manually' add them. When we implement them to the core, it will be easy to configure them, just like default `dataLabels`.

Regarding padding, this is how SVG works, see:
- textPath: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/textPath 
- text: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/text

Because that label is build as `<text> <textPath> </textPath> </text>`, we can apply a translation (`dx` and `dy`) on `<text>` element and `<textPath>` will inherit this: https://jsfiddle.net/BlackLabel/xhwstejk/30/

Regarding new line, could you show a mockup how this should behave? Thanks!

**Comment by timdisher** [comment]
> In API go to Classes tab and check SVGRenderer class: https://api.highcharts.com/class-reference/Highcharts.SVGRenderer.html#createElement The part with linkLabels is not supported (that's why this issue is marked as an enhancement ;) ) so we need to 'manually' add them. When we implement them to the core, it will be easy to configure them, just like default dataLabels.

Perfect, thank you! 

> Regarding new line, could you show a mockup how this should behave? Thanks!

Looking for behaviour as in below:
![image](https://user-images.githubusercontent.com/28613433/52124963-348d2980-2601-11e9-9742-732c91d45227.png)

The only two other features that would be great once you implement in core would be if it would be possible for the labels to try to stay upright to limit how often you have to read something upside down, and an easy way to add some extra details on mouse over. Right now I hack them in with html in node name i.e.:
`name: "<b>Treatment</b>: Placebo <br>" +
                "<b>N</b>: 6372 <br>" +
                "<b>Class</b>: NA"`

Which causes predictable errors when you try to download as SVG (attributes cannot have unescaped '<')

**Comment by TorsteinHonsi** [comment]
> this chart looks similar to the dependancy wheel - will that work?

@pawel The layout is a bit similar, but I think, at least for now, we can't make it look like this.

> should we include it in next release of network-graph?

@sebastianbochan I like the POC, but we need to use re-use dataLabel options and logic for the link labels. It should be consistent with link data labels for sankey: https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/plotoptions/sankey-link-datalabels/

**Comment by pawelfus** [comment]
@timdisher 

Simple plugin to prevent labels rendering upside down: 
```
  H.wrap(H.seriesTypes.networkgraph.prototype.pointClass.prototype, 'getLinkPath', function(proceed, from, to) {
    var left = from,
      right = to;

    if (left.plotX > right.plotX) {
      left = to;
      right = from;
    }

    return [
      'M',
      left.plotX,
      left.plotY,
      'L',
      right.plotX,
      right.plotY
    ];
  });
```
Demo: http://jsfiddle.net/BlackLabel/xhwstejk/34/

Regarding tooltip, see more details about [`tooltip.formatter`](https://api.highcharts.com/highcharts/tooltip.formatter): http://jsfiddle.net/BlackLabel/xhwstejk/36/ (Again, used custom property in `node` options which later we can use in `tooltip.formatter`).

**Comment by pawelfus** [comment]
Implemented in v7.1
